File: frontend_projects/PromptEditor/src/features/workflow/promptRaw.ts
import { ensureApiClientReady, postWorkflow } from '@/services/apiClient'
import { useEditorContextStore } from '@/features/workflow/store'
import type { EditorContext } from '@/features/workflow/store'

export type MsgRole = 'system' | 'user' | 'assistant'

export interface PromptRawMessage {
  role: MsgRole
  content: string
  source?: Record<string, any>
}

export interface PromptRawResult {
  messages: PromptRawMessage[]
}

/**
 * 组装后端 RAW 工作流输入
 * - presets: 对象（可为空对象）
 * - world_books: 数组（可空数组）
 * - history: OpenAI messages 数组（可空数组）
 * - character/persona: 若存在则传入，否则省略字段
 */
export function buildPromptRawPayload(ctx: EditorContext): Record<string, any> {
  const body: Record<string, any> = {
    presets: ctx.presets.data ?? {},
    world_books: Array.isArray(ctx.worldbook.entries) ? ctx.worldbook.entries : [],
    history: Array.isArray(ctx.history.messages) ? ctx.history.messages : [],
  }
  if (ctx.characters) body.character = ctx.characters
  if (ctx.user) body.persona = ctx.user
  return body
}

/**
 * 直接调用 RAW 工作流
 * - path: /api/workflow/smarttavern/prompt_raw/assemble_full
 */
export async function callPromptRaw(payload: Record<string, any>): Promise<PromptRawResult> {
  await ensureApiClientReady()
  const res = await postWorkflow('smarttavern/prompt_raw/assemble_full', payload)
  // 结构保障
  if (!res || typeof res !== 'object' || !Array.isArray(res.messages)) {
    throw new Error('后端返回格式不正确：缺少 messages 数组')
  }
  return { messages: res.messages as PromptRawMessage[] }
}

/**
 * 便捷方法：读取前端聚合上下文 → 构造 payload → 调用 RAW 工作流
 */
export async function runPromptRaw(): Promise<PromptRawResult> {
  const store = useEditorContextStore()
  const ctx = store.snapshot()
  const payload = buildPromptRawPayload(ctx)
  return callPromptRaw(payload)
}

File: api/workflow/SmartTavern/prompt_raw/prompt_raw.py
"""
SmartTavern Prompt RAW Workflow Registration (prompt_raw.py)

职责（封装/注册层，仅声明 API，不包含实现逻辑）:
- 通过 @core.register_api 暴露工作流 API（斜杠路径 + JSON Schema）
- 实际实现委托给同目录的 impl.py

输入说明（JSON 对象/数组，非文件路径）:
- presets: 预设文档对象，包含 prompts 数组（参考 data/presets/Default.json 的结构）
- world_books: 世界书条目数组/嵌套数组/对象（参考 data/world_books 的结构）
- history: OpenAI Chat messages 数组（[{role, content}]），可不含 source 字段
- character: 可选，角色文档对象（若含 world_book.entries 将并入世界书）
- persona: 可选，用户画像文档对象

输出:
- messages: 完整提示词（prefix + in-chat 注入），每条结构 {role, content, source}（字段顺序: role → content → source）
"""

from typing import Any

import core  # type: ignore

from .impl import assemble_full as _assemble_full


@core.register_api(
    path="smarttavern/prompt_raw/assemble_full",
    name="提示词装配（RAW：prefix + in-chat）",
    description="聚合 framing 与 in-chat，输入为 JSON 对象/数组（非文件路径），输出完整 messages（含 source）",
    input_schema={
        "type": "object",
        "properties": {
            "presets": {"type": "object", "additionalProperties": True},
            "world_books": {"type": ["array", "object"]},
            "history": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "role": {"type": "string", "enum": ["system", "user", "assistant"]},
                        "content": {"type": "string"},
                        "source": {"type": "object", "additionalProperties": True},
                    },
                    "required": ["role", "content"],
                    "additionalProperties": True,
                },
            },
            "character": {"type": "object", "additionalProperties": True},
            "persona": {"type": "object", "additionalProperties": True},
            "variables": {"type": "object", "additionalProperties": True},
        },
        "required": ["presets", "history"],
    },
    output_schema={
        "type": "object",
        "properties": {
            "messages": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "role": {"type": "string", "enum": ["system", "user", "assistant"]},
                        "content": {"type": "string"},
                        "source": {"type": "object", "additionalProperties": True},
                    },
                    "required": ["role", "content", "source"],
                },
            }
        },
        "required": ["messages"],
    },
)
async def assemble_full(
    presets: dict[str, Any],
    history: list[dict[str, Any]],
    world_books: Any = None,
    character: dict[str, Any] | None = None,
    persona: dict[str, Any] | None = None,
    variables: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """
    适配器：转发到实现层（impl.py），遵循 “API 优先 / 解耦” 原则。
    """
    return await _assemble_full(
        presets=presets,
        history=history,
        world_books=world_books,
        character=character,
        persona=persona,
        variables=variables,
    )


if __name__ == "__main__":
    # 本文件为工作流 API 的注册封装层，非独立可执行脚本。
    # 请通过 API 网关运行或使用测试脚本进行验证：
    #   python api/workflow/smarttavern/prompt_raw/test_prompt_workflow.py
    import json

    print(
        json.dumps(
            {
                "message": "This file registers the RAW workflow API. Please run the API gateway or the test script instead."
            },
            ensure_ascii=False,
        )
    )
